Src/Support/Google.Apis.Core/Http/HttpClientFactory.cs

```diff
@@ -88,18 +91,61 @@ protected virtual HttpMessageHandler CreateHandler(CreateHttpClientArgs args)
         /// <summary>
         /// Creates a simple client handler with redirection and compression disabled.
         /// </summary>
-        private HttpClientHandler CreateAndConfigureClientHandler()
+        private HttpMessageHandler CreateAndConfigureClientHandler()
         {
             var handler = CreateClientHandler();
-            if (handler.SupportsRedirectConfiguration)
+
+            ConfigureClientHandler(GetHttpClientHandler(handler));
+
+            return handler;
+        }
+
+        /// <summary>
+        /// Configure a simple client handler with redirection and compression disabled.
+        /// </summary>
+        private void ConfigureClientHandler(HttpClientHandler httpClientHandler)
+        {
+            if (httpClientHandler.SupportsRedirectConfiguration)
             {
-                handler.AllowAutoRedirect = false;
+                httpClientHandler.AllowAutoRedirect = false;
             }
-            if (handler.SupportsAutomaticDecompression)
+
+            if (httpClientHandler.SupportsAutomaticDecompression)
             {
-                handler.AutomaticDecompression = DecompressionMethods.None;
+                httpClientHandler.AutomaticDecompression = DecompressionMethods.None;
             }
-            return handler;
+        }
+
+        private HttpClientHandler GetHttpClientHandler(HttpMessageHandler httpMessageHandler)
+        {
+            if (httpMessageHandler is HttpClientHandler httpClientHandler)
+            {
+                return httpClientHandler;
+            }
+
+            if (httpMessageHandler is DelegatingHandler delegatingHandler)
+            {
+                HttpMessageHandler targetHttpMessageHandler = null;
+                while (targetHttpMessageHandler == null)
+                {
+                    var innerHandler = delegatingHandler.InnerHandler;
+                    if (innerHandler is DelegatingHandler innerDelegatingHandler)
+                    {
+                        delegatingHandler = innerDelegatingHandler;
+                    }
+                    else
+                    {
+                        targetHttpMessageHandler = innerHandler;
+                    }
+                }
+
+                if (targetHttpMessageHandler is HttpClientHandler delegatedHttpClientHandler)
+                {
+                    return delegatedHttpClientHandler;
+                }
+            }
+
+            throw new UnableToResolveHttpClientHandlerException(httpMessageHandler);
```

Comment:
This is an "artificial" restriction on the `HttpMessageHandler` return type that you want to change `HttpClientFactory.CreateClientHandler` to. If we say we can work with `HttpMessageHandler` we should be able to work with all (at least almost all) `HttpMessageHandler` implemenations, not just 2.
We have restricted the type to `HttpClientHandler` because what we can work with is an `HttpClientHandler`, not other implementations of `HttpMessageHandler`.

Reply:
In some way, it is. But `CreateHandler` depends on `HttpClientHandler` so it can distinguish which handler to create.

Reply:
We can redesign it another way.

Create `IHttpMessageHandlerConfigurator` with methods `bool SupportsAutomaticDecompression(HttpMessageHandler)` and `void SetAutomaticDecompression(HttpMessageHandler, DecompressionMethods)`. Which will be dependency of the `HttpClientFactory`, so anyone can replace/extend it.

But that is also a BC break.